### PR TITLE
add grab-bag index file to enable desctructuring requires

### PR DESCRIPTION
**Pull request by mixmix**
@mmckegg !  here's the grab bag I really want to use.
I think this is the best way to do it.

Can you plz skim the name to check there's nothing should be private (i.e. not included in the grab-bag)

### CODE REVIEW

**Comment by mmckegg** [comment]
Can you give me an example of how you would use this in a destructuring context? I'm not sure that using the module names (e.g. `dict-to-collection`) is the best approach for the grab bag. Maybe it's class/function name should be used. For example: 

```
module.exports = {
  'MutantArray': require('./array'),
  'MutantDict': require('./dict'),
  'computed': require('./computed'), // this is one is very open to discussion!
  'concat': require('./concat'),
  'dictToCollection': require('./dict-to-collection')
}
```

Or maybe it should export both?

Any thoughts @mixmix?

**Comment by ahdinosaur** [comment]
i'm with @mmckegg, as i think @mixmix you want to change [your code here](https://github.com/ssbc/patchbay/blob/a93d4ad7bea1f6e6dcd43bcfc44addcb0d95f06a/modules_extra/network.js#L6-L12) from:

```js
const Struct = require('@mmckegg/mutant/struct')
const Value = require('@mmckegg/mutant/value')
const Dict = require('@mmckegg/mutant/dict')
const toCollection = require('@mmckegg/mutant/dict-to-collection')
const mutantMap = require('@mmckegg/mutant/map')
const when = require('@mmckegg/mutant/when')
const computed = require('@mmckegg/mutant/computed')
```

to

```js
const { Struct, Value, Dict, toCollection, mutantMap, when, computed } = require('@mmckegg/mutant')
```

(not sure how we want to prefix with `mutant` considering the confict with JavaScript `Array` and `Map`)

or the old school way:

```js
const M = require('@mmckegg/mutant')
```

**Comment by mixmix** [comment]
Yeah I agree with the proposed change. Not sure about the best way to resolve the conflict on names ... maybe : 
- MMap, MArray
- ~Map, ~Array ... mmm that woun't work

**Comment by ahdinosaur** [comment]
i'm thinking now it should be as it is, so you use it like:

```js
const { Array: MutantArray } = require('@mmckegg/mutant')

const array = MutantArray()
```

```js
const Mutant = require('@mmckegg/mutant')

const array = Mutant.Array()
```

**Comment by mixmix** [comment]
yeah I'm divided  - it's kinda setting the user up from a bad time surely ...

**Comment by mixmix** [comment]
Yeah I'm sold on the readability of what you've proposed Mikey. @mmckegg can you have a look and check I've capitalised all the constructor-type files (e.g. Map, Array,...)

**Comment by mmckegg** [comment]
It's really hard to say what the correct casing is on some of those. I would have done basically exactly what you did @mixmix in a72d8cc116aa4850939368c3969044a3dc577512, except I also would have capitalised `MappedArray`, `MappedDict` (these are hybrids of Array/Dict and map - they both store their own data).

But for a simple rule, I would propose that if the function has a `set()` function on it, then it should be capitalised.

Based on that though, we probably should make `map` camelCase since it doesn't actually host any data. (that would also avoid conflicting with the global map type)

---

On the other hand, maybe we should be capitalising anything that **returns an "observable"**. In that case the new list would also include:

- `Computed`
- `Concat`
- `DictToCollection`
- `IdleProxy`
- `Keys`
- `Lookup`
- `MappedArray`
- `MappedDict`
- `Merge`
- `ProxyCollection`
- `ProxyDict`
- `Proxy`
- `Throttle`
- `When`

So basically everything except: `forEach`, `forEachPair`, `isObservable`, `onceIdle`, `resolve`, `send`, `watchAll`, `watch`, `watchThrottle`

---

**html-element** is the big question. It constructs a dom element, that in most ways behaves like an observable in the mutant world. If nothing else, it should also get `h` as an alias.

---

**In summary: There is no easy answer.** We need to come up with something that is simple and consistent. That is what was nice about the original proposition. Just a shame about the destructuring problems.

Thoughts @mixmix @ahdinosaur ?

**Comment by ahdinosaur** [comment]
oh, hmm. i think both "if the function has a set() function on it, then it should be capitalised" and "we should be capitalising anything that returns an "observable"" are both good ideas, i just can't tell which is better. :thinking: 

regarding `html-element`, i'd do `HtmlElement` (or `HTMLElement` depending on your style of caps). but most "virtual dom" implementations use `createElement` (`react`, `virtual-dom`, etc) at `create-element.js` to mirror `document.createElement`. or, i notice in the `mutant` code it's simply called `Element`, i also like that.

**Comment by mixmix** [comment]
yes, these sound like nice documentable guides. I don't have a strong intuition about set vs. returning an observable. 

I did have a feeling that I would like these myriad files in different folders describing what they are. e.g. which of these are `constructors` versus `helpers`
I wondered if answering this might help clarify conventions (and also make it easier for someone looking inside mutant for the first time - for me I feel like I'm poking around guessing what each one is at the moment, case in point me asking which ones are constructors.)

**Comment by ahdinosaur** [comment]
1. types (observables with that store data)
2. transforms (observables that depend for data)
3. helpers (a function that does something else)
